### DISCUSSION TITLE
Invalid or unexpected token (at bundle.js:2:8077) after applying the setup guide on guides/asset-management

### DISCUSSION THREAD

**Discussion by happymalyo** [webpack/webpack#16404] [Q&A]
# Bug report
Invalid or unexpected token (at bundle.js:2:8077)

**What is the current behavior?**
webpack-demo
  |- package.json
  |- package-lock.json
  |- webpack.config.js
  |- /dist
    |- bundle.js
    |- index.html
  |- /src
   |- style.css
    |- index.js
  |- /node_modules

**in index.html**
<code>
<!doctype html>
<html>
  <code><head></code>
    <title>Assets management</title><br/>
  <code></head></code>
  <code><body>
    <script src="bundle.js"></script>
  </body>
</code>
</html>
</code>

**in webpack.config.js**
 const path = require('path');

 module.exports = {
   entry: './src/index.js',
   output: {
     filename: 'bundle.js',
     path: path.resolve(__dirname, 'dist'),
   },
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: ['style-loader', 'css-loader'],
      },
    ],
  },
 };

**If the current behavior is a bug, please provide the steps to reproduce.**
When I refresh my browser, the bundle.js error appear

**What is the expected behavior?**
The command line 'npm run webpack' has no problem. But bundle.js gets error after refreshing browser like: *Invalid or unexpected token (at bundle.js:2:8077)*

**Other relevant information:**
webpack: "^5.74.0",
webpack-cli: "^4.10.0"
Node.js version: 16.16.0
Operating System: ubuntu 16.04
Additional tools: css-loader : "^6.7.1", style-loader : "^3.3.1",

**Comment by alexander-akait** [comment]
Please provide example of the problem, `npm run webpack` is not help to undestand - provide a configuraion and example of code

**Comment by foyzhao** [comment]
This scenario is very rare, and it is impossible to upload a large project as an example. The generated bundle JS is not in valid js format and browser print "Uncaught SyntaxError: Invalid or unexpected token"

Below is the snippet of the output code:
```
{case 0:return n=t.business,r=t.model,i=t.modelIds,a=void 0===i?[]:i,o=t.fieldNames,s=void 0===o?
[]:o,e.abrupt("return",this.get("/export/fields/batch/run-by-name",{business:n,model:r,modelIds:a,fieldNames:s}));
case 2:case"end":return e.stop()}}),e,this)}))),function(e){return t.apply(this,arguments)})}]),h}(c.b))||r;
t.a=new p},48ba:function(e,t,n){"use strict";n("ff48")},fd6c:function(e,t,n){"use strict";n.d(t,"b",(function(){return a})),
n("99af");var r=n("ed08"),i=n("8165"),a=function(){return{provide:i.provide,inject:i.inject,provideKey:function(e,t){
var n=t||"".concat(e||"","(").concat(Object(r.g)(),")");return Object(i.provide)(e,n),n},injectKey:function(e){var
```
Note that there is a variable name in the middle that starts with a number **48ba:function(** 

Same problem here [[Invalid JS in Webpack 5 bundle](https://stackoverflow.com/questions/71142869/invalid-js-in-webpack-5-bundle-some-numbers-from-source-code-are-replaced-by-nu)](https://stackoverflow.com/questions/71142869/invalid-js-in-webpack-5-bundle-some-numbers-from-source-code-are-replaced-by-nu)

**Reply by alexander-akait** [reply]
it can happen due terser bugs, please check you have the latest version

**Reply by foyzhao** [reply]
At present, yes.
webpack@5.75.0 -> terser-webpack-plugin@5.3.6 -> terser@5.16.0

**Reply by alexander-akait** [reply]
hm, strange, can you disable minimizer and check it (optimization.minimizer: false)?

**Reply by foyzhao** [reply]
Yes, when optimization.minimize is false, the part of the generated code is:
```
/* harmony default export */ __webpack_exports__["a"] = (new FieldRuntimeService());

/***/ }),

/***/ "48ba":
/***/ (function(module, __webpack_exports__, __nested_webpack_require_11984505__) {

"use strict";
/* harmony import */ var _node_modules_mini_css_extract_plugin_dist_loader_js_ref_9_oneOf_1_0_node_modules_css_loader_dist_cjs_js_ref_9_oneOf_1_1_node_modules_vue_loader_lib_loaders_stylePostLoader_js_node_modules_postcss_loader_src_index_js_ref_9_oneOf_1_2_node_modules_css_unicode_loader_index_js_node_modules_sass_loader_dist_cjs_js_ref_9_oneOf_1_4_node_modules_cache_loader_dist_cjs_js_ref_1_0_node_modules_vue_loader_lib_index_js_vue_loader_options_uploadFile_vue_vue_type_style_index_0_lang_scss___WEBPACK_IMPORTED_MODULE_0__ = __nested_webpack_require_11984505__("ff48");
/* harmony import */ var _node_modules_mini_css_extract_plugin_dist_loader_js_ref_9_oneOf_1_0_node_modules_css_loader_dist_cjs_js_ref_9_oneOf_1_1_node_modules_vue_loader_lib_loaders_stylePostLoader_js_node_modules_postcss_loader_src_index_js_ref_9_oneOf_1_2_node_modules_css_unicode_loader_index_js_node_modules_sass_loader_dist_cjs_js_ref_9_oneOf_1_4_node_modules_cache_loader_dist_cjs_js_ref_1_0_node_modules_vue_loader_lib_index_js_vue_loader_options_uploadFile_vue_vue_type_style_index_0_lang_scss___WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__nested_webpack_require_11984505__.n(_node_modules_mini_css_extract_plugin_dist_loader_js_ref_9_oneOf_1_0_node_modules_css_loader_dist_cjs_js_ref_9_oneOf_1_1_node_modules_vue_loader_lib_loaders_stylePostLoader_js_node_modules_postcss_loader_src_index_js_ref_9_oneOf_1_2_node_modules_css_unicode_loader_index_js_node_modules_sass_loader_dist_cjs_js_ref_9_oneOf_1_4_node_modules_cache_loader_dist_cjs_js_ref_1_0_node_modules_vue_loader_lib_index_js_vue_loader_options_uploadFile_vue_vue_type_style_index_0_lang_scss___WEBPACK_IMPORTED_MODULE_0__);
/* unused harmony reexport * */


/***/ }),

/***/ "fd6c":
/***/ (function(module, __webpack_exports__, __nested_webpack_require_11986296__) {

```
It looks like the quotes around "48ba" are removed after code minimize, If it starts with a letter like "fd6c" below, no problem, but it starts with a number, so causes a syntax error.

**Reply by foyzhao** [reply]
Another strange thing is when I add or delete some code in a file under /src, or even just modify the value of a string variable, then build again, the "48ba" in venders.[hash].js just gone, and everything is normal again. Anyway, it's a very rare case and happened a few times after I upgraded to webpack5, but didn't happen to me when I was using webpack4.
```
  output: {
    clean: true,
    publicPath: '/',
    filename: 'static/js/[name].[contenthash:4].js',
  },
  optimization: {
    splitChunks: {
      cacheGroups: {
        commons: {
          test: /[\\/]node_modules[\\/]/,
          name: 'vendors',
          chunks: 'all',
        },
      },
    },
  },
```
Modifying the code under /src cause venders.[hash].js file change, does it reasonable?

**Reply by alexander-akait** [reply]
> It looks like the quotes around "48ba" are removed after code minimize, If it starts with a letter like "fd6c" below, no problem, but it starts with a number, so causes a syntax error.

Do you mean after set `minimizer: false` you don't have problem, right?

> [contenthash:4]

I strongly recommend don't using `4`, it is very small, please use `8` as minimum

**Reply by foyzhao** [reply]
Yes, when `minimize: false` or `minimizer:[]` is set, the program works fine in the browser.

**Reply by foyzhao** [reply]
I have modified `output.filename` to `[contenthash:8]` and use it from now on. (It has no effect on the current problem, `48ba:function(e,t,n)...` is still there after build)

**Reply by foyzhao** [reply]
Ah, when I changed this to 8, the problem seemed to be solved.
```
new MiniCssExtractPlugin({
  filename: 'static/css/[name].[contenthash:8].css',
}),
```
I tried to modify some files under /src and run build, the hash of the venders.[hash].js file no longer changes.

**Reply by alexander-akait** [reply]
I think the problem is overlapping, when you have very small hash, webpack accidentally can change your code, for example:
```js
var myVar = 1234;
``` 
And when you have very small size of a hash we can change it to:
```
var myVar = 12ab; // Invalid number, but you can have `1234` contenthash in your assets
```

yeah it is very rare case, but it can happend, so my recommendation is `8`

**Reply by foyzhao** [reply]
Got it, all my projects have been changed to 8. Thanks!

**Reply by EveChee** [reply]
> I think the problem is overlapping, when you have very small hash, webpack accidentally can change your code, for example:
> 
> ```js
> var myVar = 1234;
> ```
> 
> And when you have very small size of a hash we can change it to:
> 
> ```
> var myVar = 12ab; // Invalid number, but you can have `1234` contenthash in your assets
> ```
> 
> yeah it is very rare case, but it can happend, so my recommendation is `8`

I checked part of the code of the plugin RealContentHashPlugin, and I am friends with foyzhao. So I debugged his code and got a hashMap like this
<img width="550" alt="image" src="https://user-images.githubusercontent.com/20162788/204177376-59b6c75c-2b59-4f43-a28f-95d82745eec7.png">
Then the content is replaced in the RealContentHashPlugin
<img width="602" alt="image" src="https://user-images.githubusercontent.com/20162788/204177518-b3142948-e567-4470-9f81-ff8a6c9287da.png">
<img width="581" alt="image" src="https://user-images.githubusercontent.com/20162788/204179975-f28f26c1-fb9a-4c01-a271-5ab02e8e1ec3.png">

So if it is replaced like this, there are several problems:
1.~As long as there is a hash at the beginning of a number, a syntax error will occur
2.~Related, as long as a certain js or css is changed or the vendor is changed in a passive chain, changes and new hashes will be generated.
The above is the information I got from debugging, I hope it will help to fix or troubleshoot this problem

**Reply by alexander-akait** [reply]
Yeah, it is known limitation, reparsing of generated files will be very bad for perf, also to replace content hash in the each file we need to know how to parse files (webpack works not only with JS), we have an idea - keep places of hashes (start and end index) and use it to replace (but it requires developer to change code in their loaders/plugins).
